posts/shadow-dom-and-event-propagation.md

```diff
@@ -0,0 +1,267 @@
+
+By default, events don't propagate outside shadow trees. This default behavior ensures that internal DOM events don't inadvertently leak into the document.
+
+In the example, dispatching a bubbling event from the `span#d` propagates to `p#c` and the shadow root, but it stops there. The event never reaches `div#a` since the element lives outside the shadow tree.
+
+<event-visualizer label="Single shadow tree" event-bubbles>
+  <template>
+    <div id="a">
+      <template shadowroot="open">
+        <p id="c">
+          <span id="d" target></span>
+        </p>
+        <p id="e"></p>
+      </template>
+      <img id="b"></img>
+    </div>
+  </template>
+</event-visualizer>
+
+**For an event to traverse shadow DOM boundaries, it has to be configured as a [`composed`](https://developer.mozilla.org/en-US/docs/Web/API/Event/composed)**: `new Event('test', { bubbles: true, composed: true })`. Going back to the previous example, if `composed` is enabled, you can see the event escaping the shadow tree and reaching to `div#a`.
+
+When the `composed` option is set to `true`, the dispatch event not only traverses its shadow boundary but also any other parent boundary. In the following example, the bubbling and composed event dispatch from `span#e` bubble to `div#c`, `div#a`, and their respective shadow roots.
+
+<event-visualizer label="Nested shadow trees" event-bubbles event-composed>
+  <template>
+    <div id="a">
+      <template shadowroot="open">
+        <div id="c">
+          <template shadowroot="open">
+            <span id="e" target></span>
+          </template>
+          <p id="d"></p>
+        </div>
+      </template>
+      <img id="b"></img>
+    </div>
+  </template>
+</event-visualizer>
+
+This might be counterintuitive but, a composed event always propagates outside the shadow boundary regardless if it is bubbling or not. Give a shot and set `bubbles` to false in the previous example to see how the event propagates.
+
+As you can see, when the dispatched event is composed only, the event propagates from one host element to another, `div#c` and `div#a`, without propagating through the intermediary nodes. When thinking about DOM events propagation, **`bubbles` indicates if the event propagates throw the parent hierarchy** while **`composed` indicates if the event should propagate throw the shadow DOM hierarchy**. A bubbling and composed event propagate throw all the nodes from the dispatched one up to the document root.
+
+Now that you better understand how event propagates in the shadow DOM, it is important to call out that you should think carefully about how your events are configured especially if you are building some complex applications. While it might be tempting to make all the composed and bubbling, it should not be your go-to events configuration. Events are part of the public API exposed by a web component. Not all events are equal and only certain events are worth being exposed outside the component shadow tree.
```

Comment:
```suggestion
Now that you better understand how events propagate in the shadow DOM, it is important to call out that you should think carefully about how your events are configured, especially if you are building some complex applications. While it might be tempting to make them all composed and bubbling, it should not be your go-to events configuration. Events are part of the public API exposed by a web component. Not all events are equal, and only certain events are worth being exposed outside the component shadow tree.
```